runtime/win32.c

```diff
@@ -622,20 +624,33 @@ void caml_win32_unregister_overflow_detection(void)
 
 int caml_win32_random_seed (intnat data[16])
 {
-  /* For better randomness, consider:
-     http://msdn.microsoft.com/library/en-us/seccrypto/security/rtlgenrandom.asp
-     http://blogs.msdn.com/b/michael_howard/archive/2005/01/14/353379.aspx
-  */
+  int n = 0;
+  unsigned char buffer[12];
+  int nread = 0;
+
+  /* Try kernel entropy first */
+  NTSTATUS st = BCryptGenRandom(NULL, buffer, 12,
+                                BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+  if (NT_SUCCESS(st))
+    nread = 12;
+  while (nread > 0) data[n++] = buffer[--nread];
+
+  /* If the kernel provided enough entropy, we now have 96 bits
+     of good random data and can stop here. */
+  if (n >= 12) return n;
+
+  /* Otherwise, complement whatever we got (probably nothing)
+     with some not-very-random data. */
   FILETIME t;
   LARGE_INTEGER pc;
   GetSystemTimeAsFileTime(&t);
   QueryPerformanceCounter(&pc);  /* PR#6032 */
-  data[0] = t.dwLowDateTime;
-  data[1] = t.dwHighDateTime;
-  data[2] = GetCurrentProcessId();
-  data[3] = pc.LowPart;
-  data[4] = pc.HighPart;
-  return 5;
+  if (n < 16) data[n++] = t.dwLowDateTime;
```

Comment:
`n` is always zero if you reach here, or am I missing something?

Reply:
Yes, you're right. `BCryptGenRandom` should either write 12 bytes and succeed; or fail and `n` will be zero.